Copy question link from dashboard

kamilmielnik opened the pull request:
Closes #41070

### Description

Makes titles of most dashboard charts actual HTML anchors.
Thanks to this users can right-click the title to copy URL, middle-click it to open it in a new tab, drag & drop it to browser's address bar, etc.

### How to verify

1. Create a dashboard
2. Add various visualizations to it
3. Right-click visualization title 

Context menu should open and you should be able to copy the URL of the link.

### Demo


https://github.com/metabase/metabase/assets/6830683/b31a0664-7367-4fd0-b6bc-9ce5a21eed1c

kamilmielnik committed:
Remove dead CSS

kamilmielnik committed:
Use proper types

kamilmielnik committed:
Properly type parameterValues

kamilmielnik committed:
Add href prop support to LegendCaption

kamilmielnik committed:
Extract getNewCardUrl
- move getParametersMappedToDashcard

kamilmielnik committed:
Use href in CartesianChart

kamilmielnik commented on frontend/src/metabase/css/dashboard.module.css:
```diff
@@ -198,8 +198,7 @@
 .Dashboard :global(text.value-label),
-.Dashboard :global(text.value-label-white),
-.Dashboard .LineAreaBarChart :global(.dc-chart .axis text) {
+.Dashboard :global(text.value-label-white) {
```
`.LineAreaBarChart` does not exist anymore

kamilmielnik committed:
Handle RowChart

kamilmielnik committed:
Handle Funnel

kamilmielnik committed:
Format code

kamilmielnik committed:
Fix types

kamilmielnik committed:
Merge branch 'master' into 41070-dashboard-question-link

kamilmielnik committed:
Update first test in title-drill

kamilmielnik committed:
Update test

kamilmielnik committed:
Update test

kamilmielnik committed:
Update tests
- pass questionId in createQuestionAndDashboard

kamilmielnik commented on e2e/test/scenarios/dashboard/title-drill.cy.spec.js:
```diff
@@ -28,30 +29,32 @@ describe("scenarios > dashboard > title drill", () => {
+
+          getDashboardCard().findByRole("link", { name: "Q1" }).as("title");
+          cy.get("@title")
+            .should("have.attr", "href")
+            .and("include", `/question/${questionId}`);
```
These are new link-related assertions

kamilmielnik commented on e2e/test/scenarios/dashboard/title-drill.cy.spec.js:
```diff
@@ -28,30 +29,32 @@ describe("scenarios > dashboard > title drill", () => {
+          cy.get("@title").click();
+
+          queryBuilderMain().within(() => {
+            cy.findByText("This question is written in SQL.").should("exist");
+            cy.findByText("foo").should("exist");
+            cy.findByText("bar").should("exist");
```
I kept the old assertions - maybe they are useful.

kamilmielnik commented on frontend/src/metabase/dashboard/actions/getNewCardUrl.ts:
```diff
@@ -0,0 +1,110 @@
+ * should come from this function, which merges any currently applied dashboard filters
+ * into the new card / URL parameters.
+ */
+export const getNewCardUrl = (
```
This function has been extracted from `navigateToNewCardFromDashboard` with only minor changes

kamilmielnik commented on frontend/src/metabase/dashboard/actions/getNewCardUrl.ts:
```diff
@@ -0,0 +1,110 @@
+  return url;
+};
+
+export function getParametersMappedToDashcard(
```
Moved from `frontend/src/metabase/parameters/utils/dashboards.ts`.

kamilmielnik commented on frontend/src/metabase/dashboard/actions/getNewCardUrl.ts:
```diff
@@ -0,0 +1,110 @@
+};
+
+export function getParametersMappedToDashcard(
+  parameters: Dashboard["parameters"],
```
We used to pass an entire `Dashboard` here, but only parameters are needed.
I had do to this change due to `StoreDashboard` not being the same as `Dashboard`.

kamilmielnik commented on frontend/src/metabase/dashboard/actions/getNewCardUrl.unit.spec.ts:
```diff
@@ -0,0 +1,69 @@
+import {
```
Tests moved from `frontend/src/metabase/parameters/utils/dashboards.unit.spec.js` and updated to use TS.

kamilmielnik commented on frontend/src/metabase/parameters/utils/dashboards.ts:
```diff
@@ -210,28 +209,6 @@ function buildFieldFilterUiParameter(
   };
 }
 
-export function getParametersMappedToDashcard(
```
Moved to `getNewUrl.ts`

kamilmielnik commented on frontend/src/metabase/parameters/utils/dashboards.unit.spec.js:
```diff
@@ -191,65 +190,6 @@ describe("metabase/parameters/utils/dashboards", () => {
     });
   });
 
-  describe("getParametersMappedToDashcard", () => {
```
Tests moved to `frontend/src/metabase/dashboard/actions/getNewCardUrl.unit.spec.ts`

kamilmielnik committed:
Add test for links

kamilmielnik committed:
Merge branch 'master' into 41070-dashboard-question-link

kamilmielnik committed:
Merge branch 'master' into 41070-dashboard-question-link

kamilmielnik committed:
Fix createDashboardWithQuestions

kamilmielnik committed:
Add missing click handler

kamilmielnik committed:
Fix crash

kamilmielnik committed:
Fix typo

kamilmielnik committed:
Add a workaround for #42999

kulyk reviewed (commented):
Looks great, and I love the change. The PR comments were very helpful too, thanks 🙏 

@kamilmielnik I noticed two minor things that we might want to address before merging:

1. it seems dashcard titles no longer have a hover style? They used to change color to "brand" on hover, but now only the cursor is changing
2. clicking a title with CMD key pressed opens two question tabs for me

(I'm using Arc browser fwiw. I couldn't reproduce the 2nd thing on Chrome, but reproed in Firefox)

kamilmielnik committed:
Merge branch 'master' into 41070-dashboard-question-link

kamilmielnik committed:
Fix tests

kamilmielnik committed:
Fix hover color

kamilmielnik committed:
Update comment

kamilmielnik commented:
> 1. it seems dashcard titles no longer have a hover style? They used to change color to "brand" on hover, but now only the cursor is changing

Oh, great catch! :heart: :eagle: :eye: 
This got broken after merging recent `master` because a CSS variable has been renamed.
Fixed: 30e37bef02acb83d8e125f9c019184ebcbbdf419

> 2. clicking a title with CMD key pressed opens two question tabs for me

Fixed with f9cfee783eca20d10512283eef7ffe225606ad34 + c29da1288bdbba58982df623973e7ea89f420a5.

kamilmielnik committed:
Merge branch 'master' into 41070-dashboard-question-link

ranquild commented on frontend/src/metabase/dashboard/actions/getNewCardUrl.ts:
```diff
@@ -0,0 +1,124 @@
+ */
+export const getNewCardUrl = (
+  metadata: Metadata,
+  dashboardState: DashboardState,
```
Please accept only parameters that you need, e.g. dashboardId etc

kamilmielnik replied:
Done: b804da443bc9d461b3ab1e064fe98c856c96c3e8

ranquild reviewed (commented):
Can we change `getNewCardUrl` (and props drilling) to 1) connecting as close as possible to the actual Visualization 2) to just url and not callback?

kamilmielnik committed:
Get rid of mapStateToProps in DashboardGrid

kamilmielnik commented on frontend/src/metabase/dashboard/components/DashboardGrid.tsx:
```diff
@@ -104,6 +108,9 @@ interface DashboardGridState {
 
+const mapStateToProps = (state: State) => ({
+  dashboardState: state.dashboard,
+});
```
It was possible to avoid `mapStateToProps`: b804da443bc9d461b3ab1e064fe98c856c96c3e8 :tada:

uladzimirdev commented on frontend/src/metabase/dashboard/components/DashboardGrid.tsx:
```diff
@@ -500,6 +503,34 @@ class DashboardGrid extends Component<DashboardGridProps, DashboardGridState> {
+    dashcard: DashboardCard;
+    objectId?: number | string;
+  }) => {
+    if (isActionDashCard(dashcard) || isVirtualDashCard(dashcard)) {
```
nit: probably `!isQuestionDashCard(dashcard)` could work

kamilmielnik replied:
8861e26cefc67592bf9a08a99d44f731ebd40b53

uladzimirdev commented on frontend/src/metabase/dashboard/components/DashCard/DashCard.tsx:
```diff
@@ -75,6 +76,9 @@ export interface DashCardProps {
   onReplaceCard: () => void;
   onRemove: () => void;
   markNewCardSeen: (dashcardId: DashCardId) => void;
+  getNewCardUrl: (
```
nit: I'd use `typeof navigateToNewCardFromDashboard` if possible

kamilmielnik replied:
Return type differs

uladzimirdev commented on frontend/src/metabase/dashboard/components/DashCard/DashCard.tsx:
```diff
@@ -236,26 +241,36 @@ function DashCardInner({
+        return null;
+      }
+
+      return ({ nextCard, previousCard, objectId }) => {
```
honestly it looks like a `useCallback` + conditional passing handler: `navigateToNewCardFromDashboard ? changeCardAndRunHandler: null)`

kamilmielnik replied:
18709164627f6f94cb721e01ac818dc295b7729f

uladzimirdev reviewed (commented):
LGTM, but please check comments

kamilmielnik committed:
Invert condition

uladzimirdev commented:
@kamilmielnik, mostly questions:

https://github.com/metabase/metabase/assets/125459446/2557e181-1e3a-48da-85ef-c8e9bf59897f

I didn't get in the details, but can't we open a saved question by id? like `/question/126` will redirect me to the correct question


if dashboard has some filters applied, should we open a question with the same filters applied?

https://github.com/metabase/metabase/assets/125459446/287039f0-899e-411f-b530-cf0d86831e32

kamilmielnik committed:
Use useCallback instead useMemo

kamilmielnik commented:
@uladzimirdev 

> if dashboard has some filters applied, should we open a question with the same filters applied?

Yes.

> I didn't get in the details, but can't we open a saved question by id? like /question/126 will redirect me to the correct question

I think the logic in `getNewCardUrl` could be improved to generate simpler URLs if possible but it's definitely out of scope of this PR. Luckily, the URL will be automatically adjusted upon navigation, so it's not that big of a deal.

kamilmielnik committed:
Do not pass getNewCardUrl down if navigateToNewCardFromDashboard is not present
- Remove default noop for navigateToNewCardFromDashboard in PublicDashboard

kamilmielnik commented:
@ranquild 

> Can we change `getNewCardUrl` (and props drilling) to 1) connecting as close as possible to the actual Visualization 2) to just url and not callback?

Unfortunately this needs to follow the same logic as `onChangeCardAndRun` which comes from the very top of components tree and is drilled down quite deeply. Presence of this prop determines whether the link can be clickable (see [DashboardData](https://github.com/metabase/metabase/blob/0e157dc8fbf94f59c5243d261865ecfb37c867c0/frontend/src/metabase/dashboard/hoc/DashboardData.jsx#L125) and [PublicDashboard](https://github.com/metabase/metabase/blob/0e157dc8fbf94f59c5243d261865ecfb37c867c0/frontend/src/metabase/public/containers/PublicDashboard/PublicDashboard.tsx#L275)).

This comment made me realize it was not properly handled before, fixed: 717d095425ef859856573c9d9cc1b69c0ba01e4b

If we were to refactor this, both `onChangeCardAndRun` and `getNewCardUrl` should get the same treatment.
I think it's out of scope of this PR though.

When it comes to passing down URL instead of a callback, we'd have to lift up (or duplicate) the logic of [transforming](https://github.com/metabase/metabase/blob/632236d3cb5b175e27c2ac5d7f89170f096e52fb/frontend/src/metabase/visualizations/components/Visualization/Visualization.jsx#L143) raw series up from `Visualization` component, which I think we should not do.

kulyk commented on frontend/src/metabase/public/containers/PublicDashboard/PublicDashboard.tsx:
```diff
@@ -272,7 +272,7 @@ class PublicDashboardInner extends Component<PublicDashboardProps> {
                   mode={PublicMode as unknown as Mode}
                   metadata={this.props.metadata}
-                  navigateToNewCardFromDashboard={() => {}}
+                  navigateToNewCardFromDashboard={undefined}
```
Can we just skip the prop altogether? 🤔

kamilmielnik replied:
7e9777b5f6e3adb834e3861763226c7c1e918675

kulyk reviewed (approved):
LGTM

kamilmielnik committed:
Remove explicit undefined prop

kamilmielnik committed:
Merge branch 'master' into 41070-dashboard-question-link

kamilmielnik committed:
Merge branch 'master' into 41070-dashboard-question-link

uladzimirdev reviewed (approved):
:shipit:
